File: src/main/kotlin/sunsetsatellite/vm/sunlite/SLNativeFunction.kt
package sunsetsatellite.vm.sunlite

import sunsetsatellite.lang.sunlite.Type

abstract class SLNativeFunction(val name: String, val returnType: Type, val arity: Int = 0) {

	override fun toString(): String {
		return "<native fn '${name}'>"
	}

	abstract fun call(vm: VM, args: Array<AnySLValue>): AnySLValue
}

File: src/main/csharp/SunliteSharp/Runtime/SLNativeFunction.cs
namespace SunliteSharp.Runtime;

using Type = SunliteSharp.Core.Type;

public record SLNativeFunction(string Name, Type ReturnType, int Arity, Func<VM, AnySLValue[], AnySLValue> Call)
{
    public override string ToString()
    {
        return $"<native fn '{Name}'>";
    }
}